File: JavaScript/25_Dynamic Programming/H-TUF-0 and 1 Knapsack.js
/**
 * TUF: 0/1 Knapsack Problem
 * https://takeuforward.org/plus/dsa/problems/0-1-knapsack
 *
 * Given weights and values of n items, put items in a knapsack of capacity W
 * to get the maximum total value.
 *
 * Type             | Details
 * -----------------+----------------------------------------------------
 * Time Complexity  | O(N * W) – DP table
 * Space Complexity | O(W) – Space optimized to 1D array
 */

/**
 * 0/1 Knapsack using 2D DP
 * @param {number} W - Knapsack capacity
 * @param {number[]} wt - Array of weights
 * @param {number[]} val - Array of values
 * @param {number} n - Number of items
 * @returns {number} - Maximum value
 */
function knapsack2D(W, wt, val, n) {
    // dp[i][w] = max value using first i items with capacity w
    const dp = Array.from({ length: n + 1 }, () => new Array(W + 1).fill(0));
    
    for (let i = 1; i <= n; i++) {
        for (let w = 0; w <= W; w++) {
            // Don't take item i
            dp[i][w] = dp[i - 1][w];
            
            // Take item i (if weight allows)
            if (wt[i - 1] <= w) {
                dp[i][w] = Math.max(dp[i][w], val[i - 1] + dp[i - 1][w - wt[i - 1]]);
            }
        }
    }
    
    return dp[n][W];
}

/**
 * 0/1 Knapsack using 1D DP (Space Optimized)
 * @param {number} W - Knapsack capacity
 * @param {number[]} wt - Array of weights
 * @param {number[]} val - Array of values
 * @param {number} n - Number of items
 * @returns {number} - Maximum value
 */
function knapsack(W, wt, val, n) {
    const dp = new Array(W + 1).fill(0);
    
    for (let i = 0; i < n; i++) {
        // Traverse from right to left to avoid using updated values
        for (let w = W; w >= wt[i]; w--) {
            dp[w] = Math.max(dp[w], val[i] + dp[w - wt[i]]);
        }
    }
    
    return dp[W];
}

/**
 * 0/1 Knapsack with item selection
 */
function knapsackWithItems(W, wt, val, n) {
    const dp = Array.from({ length: n + 1 }, () => new Array(W + 1).fill(0));
    
    for (let i = 1; i <= n; i++) {
        for (let w = 0; w <= W; w++) {
            dp[i][w] = dp[i - 1][w];
            if (wt[i - 1] <= w) {
                dp[i][w] = Math.max(dp[i][w], val[i - 1] + dp[i - 1][w - wt[i - 1]]);
            }
        }
    }
    
    // Backtrack to find selected items
    const selectedItems = [];
    let w = W;
    for (let i = n; i > 0 && w > 0; i--) {
        if (dp[i][w] !== dp[i - 1][w]) {
            selectedItems.push(i - 1); // 0-indexed
            w -= wt[i - 1];
        }
    }
    
    return { maxValue: dp[n][W], selectedItems: selectedItems.reverse() };
}

/**
 * Recursive with Memoization
 */
function knapsackMemo(W, wt, val, n) {
    const memo = new Map();
    
    function solve(i, w) {
        if (i === 0 || w === 0) return 0;
        
        const key = `${i},${w}`;
        if (memo.has(key)) return memo.get(key);
        
        let result = solve(i - 1, w); // Don't take
        if (wt[i - 1] <= w) {
            result = Math.max(result, val[i - 1] + solve(i - 1, w - wt[i - 1]));
        }
        
        memo.set(key, result);
        return result;
    }
    
    return solve(n, W);
}

// Test cases
const val = [60, 100, 120];
const wt = [10, 20, 30];
const W = 50;
const n = val.length;

console.log("Knapsack 2D:", knapsack2D(W, wt, val, n)); // Expected: 220
console.log("Knapsack 1D:", knapsack(W, wt, val, n)); // Expected: 220
console.log("Knapsack Memo:", knapsackMemo(W, wt, val, n)); // Expected: 220

const result = knapsackWithItems(W, wt, val, n);
console.log("Max Value:", result.maxValue);
console.log("Selected Items (indices):", result.selectedItems);

File: Java/25_Dynamic Programming/H-TUF-0 and 1 Knapsack.java
// 0 and 1 Knapsack
// https://takeuforward.org/plus/dsa/problems/0-and-1-knapsack

/**
 * Given two integer arrays, val and wt, each of size N, 
 * which represent the values and weights of N items respectively, 
 * and an integer W representing the maximum capacity of a knapsack, 
 * determine the maximum value achievable by selecting a subset of the items 
 * such that the total weight of the selected items does not exceed the knapsack capacity W.

 * Each item can either be picked in its entirety or not picked at all (0-1 property). 
 * The goal is to maximize the sum of the values of the selected items while keeping the total weight within the knapsack's capacity.
 * 
 * Examples:
 * Input: val = [60, 100, 120], wt = [10, 20, 30], W = 50
 * Output: 220
 * Explanation: Select items with weights 20 and 30 for a total value of 100 + 120 = 220.
 * 
 * Input: val = [10, 40, 30, 50], wt = [5, 4, 6, 3], W = 10
 * Output: 90
 * Explanation: Select items with weights 4 and 3 for a total value of 40 + 50 = 90.
 * 
 * Input: val = [20, 5, 10, 40, 15, 25], wt = [1, 2, 3, 8, 7, 4], W = 10
 * Output: 65
 * Explanation: Select items with weights 1, 3, and 4 for a total value of 20 + 10 + 35 = 65.
 */


class Solution {
    public int knapsack01(int[] wt, int[] val, int n, int W) {
        // dp[i] will store the maximum value that can be obtained with capacity i
        int[] dp = new int[W + 1];
        
        // Initialize dp array with 0 (base case: 0 value for 0 capacity)
        // Java initializes int arrays to 0 by default, so no need to explicitly set
        
        // Iterate through each item
        for (int i = 0; i < n; i++) {
            // Process the dp array from right to left to avoid overwriting values
            // that are yet to be processed in the current iteration
            for (int w = W; w >= wt[i]; w--) {
                // For each weight, decide whether to include the current item or not
                // dp[w] represents the maximum value without including the current item
                // dp[w - wt[i]] + val[i] represents the maximum value including the current item
                dp[w] = Math.max(dp[w], dp[w - wt[i]] + val[i]);
            }
        }
        
        // The answer is the maximum value that can be obtained with the given capacity W
        return dp[W];
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Test case 1
        int[] val1 = {60, 100, 120};
        int[] wt1 = {10, 20, 30};
        int W1 = 50;
        System.out.println("Test Case 1: " + solution.knapsack01(wt1, val1, val1.length, W1)); // Expected: 220
        
        // Test case 2
        int[] val2 = {10, 40, 30, 50};
        int[] wt2 = {5, 4, 6, 3};
        int W2 = 10;
        System.out.println("Test Case 2: " + solution.knapsack01(wt2, val2, val2.length, W2)); // Expected: 90
        
        // Test case 3
        int[] val3 = {20, 5, 10, 40, 15, 25};
        int[] wt3 = {1, 2, 3, 8, 7, 4};
        int W3 = 10;
        System.out.println("Test Case 3: " + solution.knapsack01(wt3, val3, val3.length, W3)); // Expected: 65
        
        // Test case 4: Edge case with zero capacity
        int[] val4 = {1, 2, 3};
        int[] wt4 = {4, 5, 1};
        int W4 = 0;
        System.out.println("Test Case 4: " + solution.knapsack01(wt4, val4, val4.length, W4)); // Expected: 0
    }
}